### PR TITLE
Restrict Un-Accredited Providers from Creating New Trainees

### PR DESCRIPTION
**Pull request by adamjcolvin**
### Context
[Trello](https://trello.com/c/HQ6tLiTf)

While un-accredited Providers should be able to edit existing Trainees (as they could as a normal Provider), we need to prevent them from creating new Trainees.

### Changes proposed in this pull request
This adds a new method to UserWithOrganisationContext called `accredited_provider`, which will check if the user’s provider has `accredited` set to true. We then check this in the `new?` and `create?` methods in the trainee policy, since only accredited providers should now be able to do this.

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?
- [ ] Do we need to send any updates to DQT as part of the work in this PR?
- [ ] Does this PR need an ADR?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

File: app/policies/trainee_policy.rb [hunk 1]
```diff
@@ -56,11 +56,11 @@ def show?
   end
 
   def new?
-    user.provider? && !user_is_read_only?
+    user.accredited_provider? && !user_is_read_only?
   end
 
   def create?
-    user_in_provider_context? && !user_is_read_only?
+    (user_in_provider_context? && user.accredited_provider?) && !user_is_read_only?
```
[the remark this answers is not in the document]
Comment (adamjcolvin) [reply app/policies/trainee_policy.rb:63]: @tomtrentham can you confirm what the logic should be here? We check that the trainee's provider is the same as the current user's provider with `user_in_provider_context?`: 

```ruby
def user_in_provider_context?
  user&.organisation == trainee.provider
end
```
But could that be true while the trainee was registered by another provider? And if so, do we allow editing / updating in that case?

[the remark this answers is not in the document]
Comment (tomtrentham) [reply app/policies/trainee_policy.rb:63]: @adamjcolvin I think there are a number of questions here which I'll try to answer:
1. Should this also apply to `update`?
   * No, we want Providers to continue to be able to update Trainees where they were the accredited Provider for that Trainee
1. Should this also apply to all the actions in cases where the trainee was registered by a different accrediting provider?
   * Trainees should only be editable by a user who's signed in as the Provider who's set as the accrediting provider for that Trainee
1. We check that the trainee's provider is the same as the current user's provider but could that be true while the trainee was registered by another provider?
   * The only scenario I can think of that matches this is for future HESA imports where the LP can create Trainees for an AP. However, we will need to [change those records so that the correct AP is set](https://trello.com/c/2A6RBm0n/1899-spike-investigate-whether-we-can-determine-the-accredited-provider-for-hesa-submissions-made-by-lead-partner-heis).

[the remark this answers is not in the document]
Comment (tomtrentham) [reply app/policies/trainee_policy.rb:63]: Yes, @adamjcolvin I think that's right (assuming you meant `new` and `create` policies).
My only other thought is whether this will impact the `TrnSubmissionsController` for unaccredited Providers in a way that we don't want?

[the remark this answers is not in the document]
Comment (adamjcolvin) [reply app/policies/trainee_policy.rb:63]: @tomtrentham I don't think there'll be any issues with `TrnSubmissionsController` after having a look. We don't ask any questions about the provider there and the trainee has already been created at that point. So an unaccredited provider won't reach that point now.

[the remark this answers is not in the document]
Comment (tomtrentham) [reply app/policies/trainee_policy.rb:63]: @adamjcolvin My concern was that an unaccredited Provider might not be able to request a TRN for an existing Trainee. This might be a real edge case but maybe adding a test in `spec/controllers/trn_submissions_controller_spec.rb` for an unaccredited provider would clarify the situation?
